# L3 Q8. Visual Assessment
This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below.

## Gather

In [97]:
import pandas as pd


In [98]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')


## Assess

In [23]:
# Visual assessment
patients


patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [7]:
treatments


given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [8]:
adverse_reactions


given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

## Conclusions

### Quality
- Patients table: zip_code data type: float not string (validity)
- Patients table: zip_code has comma (validity)
- Patients table: Tim Neudorf height is 27 instead of 72 in (validity)
- Patients table: Inconsistent data in state, full name / Short for New York (consistency)
- Patients table: NaN values in hba1c_change (completeness)
- Patients table: given_name for patient_id 9 (accuracy)
- Treatment table: Missing rows (completeness)
- Treatment table: Format u next to dose (validity)
- Treatment and Adverse_Reactions table: Lowercase name (concistency)
- All tables: The column name given_name could be more explanatory, as first_name (validity)

### Tidyness
- Treatment messy - in two different tables (but can be good for analysis) 

## Data Quality Dimensions
1. **Completeness**: Do we have all of the records that we should? Do we have missing records or not? Are there specific rows, columns, or cells missing?
2. **Validity**: We have the records, but they're not valid, i.e., they don't conform to a defined schema. A schema is a defined set of rules for data. These rules can be real-world constraints (e.g. negative height is impossible) and table-specific constraints (e.g. unique key constraints in tables).
3. **Accuracy**: Inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect. Example: a patient's weight that is 5 lbs too heavy because the scale was faulty.
4. **Consistency**: Inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing. Consistency, i.e., a standard format, in columns that represent the same data across tables and/or within tables is desired.

# L3 Q14 Programatic Assessment
 These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)




In [31]:
patients.head()


patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [32]:
patients.tail()

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
498       2530 Victoria Court   Milton Mills        ME    3852.0   
499  494 Clarksburg Park Road         Sedona        AZ   86341.0   
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [37]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [34]:
treatments.sample(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
167     lamara    dratchev  49u - 58u          -         7.75       7.40   
234     haylom       nebay          -  42u - 44u         7.62       7.22   
261   caroline      shuler          -  50u - 54u         7.63       7.27   
128      david  gustafsson          -  33u - 34u         7.72       7.28   
76    noriyuki       sakai          -  32u - 31u         7.58       7.16   

     hba1c_change  
167          0.35  
234          0.90  
261           NaN  
128          0.94  
76           0.92

In [35]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [39]:
# Nr of patients in New York City
(patients.city == 'New York').sum()

18

In [89]:
patients.surname.value_counts()
patients.patient_id[patients.surname == 'Doe']

215    216
229    230
237    238
244    245
251    252
277    278
Name: patient_id, dtype: int64

In [80]:
patients.address.value_counts()

123 Main Street          6
2476 Fulton Street       2
2778 North Avenue        2
648 Old Dear Lane        2
2645 Moore Avenue        1
                        ..
1251 Clarence Court      1
4040 Linda Street        1
4277 Mutton Town Road    1
206 Eagle Lane           1
3781 Hamill Avenue       1
Name: address, Length: 483, dtype: int64

In [87]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [90]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [91]:
sum(treatments.auralin.isnull())

0

In [92]:
sum(treatments.novodra.isnull())

0

In [ ]:
### Quality
- Patients table: zip_code data type: float not string (validity)
- Patients table: zip_code has comma (validity)
- Patients table: Tim Neudorf height is 27 instead of 72 in (validity)
- Patients table: Inconsistent data in state, full name / Short for New York (consistency)
- Patients table: NaN values in hba1c_change (completeness)
- Patients table: given_name for patient_id 9 (accuracy)
- Treatment table: Missing rows (completeness)
- Treatment table: Format u next to dose (validity)
- Treatment and Adverse_Reactions table: Lowercase name (concistency)
- All tables: The column name given_name could be more explanatory, as first_name (validity)
- Patient and Treatment table: Erroneous datatype (assigned_sex, state, zip_code, birthdate, auralin and novodra columns)
- Treatment table: Inaccurate HbA1c changes 4 mistaken as 9. (accuracy)
- Patient table: Multiple phonenr formats
- Patients table: Default John Doe data
- Patient table: Duplicate records for Jakobsen, Gersten, Taylor
- Patient table: Inconcistency in weights for weight 48.8
- Patients table: auralin and novodra, nulls represented as dashes
    
### Tidyness
- Patients table: Email and phonenr records in the same column
- Treatment table: auralin and novodra records ztwo columns in one
- Treatment table: Column headers auralin and novodra are actually variables, create a treatment column


# L4 Q1. Programmatic Data Cleaning
- Define
- Code 
- Test

In [ ]:
# Ex (you can also perform these steps for every issue to adjust)
# gather
import pandas as pd 
df = pd.read_csv('animals.csv')

# assess
df.head() 

# clean
df_clean = df.copy()

# Remove 'bb' before every animal name using string slicing
df_clean['Animal'] = df_clean['Animal'].str[2:]
# Replace ! with . in body weight and brain weight columns
df_clean['Body weight (kg)'] = df_clean['Body weight (kg)'].str.replace('!', '.')
df_clean['Brain weight (g)'] = df_clean['Brain weight (g)'].str.replace('!', '.')

# test
df_clean.head()


## Cleaning Order
1. Address Missing Data
2. Address Tidyness Issues
3. Address Quality Issues

The very first thing to do before any cleaning occurs is to make a copy of each piece of data. All of the cleaning operations will be conducted on this copy so you can still view the original dirty and/or messy dataset later.

_df_clean = df.copy()_ Copies the df and locate it to another memory location - Perform the cleaning on the copy peacefully.

In [103]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### 1. Missing Data
Do first always. 

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

#### Define
Find the missing records and append to the treatments table.

#### Code

In [105]:
# find the extra treatments
df_extra = pd.read_csv('treatments_cut.csv')

In [106]:
df_extra.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [112]:
# combine the two lists
treatments_clean = treatments_clean.append(df_extra)
# alt: treatments_clean = pd.concat([treatments_clean, treatments_cut],
                            # ignore_index=True)

#### Test

In [111]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [110]:
df_extra.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [113]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

#### Define
Calculate the correct HbA1c chance for all observation programmatically. 

#### Code

In [115]:
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end


#### Test

In [117]:
treatments_clean

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0    veronika    jindrová  41u - 48u          -         7.63       7.20   
1      elliot  richardson          -  40u - 45u         7.56       7.09   
2    yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3        skye  gormanston  33u - 36u          -         7.97       7.62   
4      alissa      montez          -  33u - 29u         7.78       7.46   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0           0.43  
1           0.47  
2           0.43  
3           0.35  
4           0.32  
..           ...  
65          0.34  
66          0.45  
67          0.30  
68          0.47  
69          0.46  

[350 rows x 7 columns]

In [118]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


### Tidyness



#### Contact column in `patients` table contains two variables: phone number and email

In [119]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

#### Code

#### Test

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

#### Define
1. Create a new column from the column names by using pd.melt function.

#### Code

In [121]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [236]:
# create treatment column
test = pd.melt(treatments_clean, id_vars = ['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], value_vars = ['auralin', 'novodra'], var_name = 'treatment', value_name = 'dose', ignore_index = False)
test.dtypes


given_name       object
surname          object
hba1c_start     float64
hba1c_end       float64
hba1c_change    float64
treatment        object
dose             object
dtype: object

In [237]:
# Move columns
tret_col = test.pop('treatment') # Remove column and save in a variable

In [238]:
d_col = test.pop('dose')

In [239]:
# Insert columns on the right place
test.insert(2, 'treatment', tret_col)

In [240]:
test.insert(3, 'dose', sd_col)

In [241]:
# Split start and end dose and insert it into df
test.insert(3, 'start_dose', test.dose.str.split('-', expand = True)[0])
test.insert(4, 'end_dose', test.dose.str.split('-', expand = True)[1])


In [243]:
# Remove dose
test.drop('dose', axis=1, inplace=True)

In [249]:
# Replace u with ''
test.start_dose = test.start_dose.str.replace('u', '', regex = False)
test.end_dose = test.end_dose.str.replace('u', '', regex = False)


In [267]:
# Change data type from object to str
test.start_dose.astype(str)
test.end_dose.astype(str)

0     48
1       
2       
3     36
4       
      ..
65      
66    26
67      
68    44
69      
Name: end_dose, Length: 700, dtype: object

In [280]:
# Change data type from str to int
test.start_dose[test.start_dose == ''].drop(axis = 1, columns = 'start_dose')

1      
2      
4      
5      
8      
     ..
62     
63     
65     
67     
69     
Name: start_dose, Length: 350, dtype: object

In [281]:
test.start_dose

0     41
1       
2       
3     33
4       
      ..
65      
66    28
67      
68    42
69      
Name: start_dose, Length: 700, dtype: object

#### Test

In [258]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 700 entries, 0 to 69
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    700 non-null    object 
 1   surname       700 non-null    object 
 2   treatment     700 non-null    object 
 3   start_dose    700 non-null    object 
 4   end_dose      700 non-null    object 
 5   hba1c_start   700 non-null    float64
 6   hba1c_end     700 non-null    float64
 7   hba1c_change  700 non-null    float64
dtypes: float64(3), object(5)
memory usage: 49.2+ KB


In [262]:
test.head()

given_name     surname treatment start_dose end_dose  hba1c_start  \
0   veronika    jindrová   auralin         41       48         7.63   
1     elliot  richardson   auralin                             7.56   
2   yukitaka    takenaka   auralin                             7.68   
3       skye  gormanston   auralin         33       36         7.97   
4     alissa      montez   auralin                             7.78   

   hba1c_end  hba1c_change  
0       7.20          0.43  
1       7.09          0.47  
2       7.25          0.43  
3       7.62          0.35  
4       7.46          0.32